lib/rom/command.rb

```diff
@@ -20,36 +20,31 @@ module ROM
   #
   # @private
   class Command
+    extend Dry::Initializer::Mixin
     include Dry::Equalizer(:relation, :options)
     include Commands
     include Pipeline::Operator
 
     extend ClassMacros
     extend ClassInterface
 
-    include Options
-
     defines :adapter, :relation, :result, :input, :register_as, :restrictable
 
-    option :type, allow: [:create, :update, :delete]
-    option :source, reader: true
-    option :result, reader: true, allow: [:one, :many]
+    # @attr_reader [Relation] relation The command's relation
+    param :relation
+
+    CommandType = Dry::Types['strict.symbol'].enum(:create, :update, :delete)
+    Result = Dry::Types['strict.symbol'].enum(:one, :many)
+
+    option :type, type: CommandType, optional: true
+    option :source, reader: true, optional: true, default: -> c { c.relation }
+    option :result, reader: true, type: Result
     option :input, reader: true
-    option :curry_args, type: Array, reader: true, default: EMPTY_ARRAY
+    option :curry_args, reader: true, default: -> _ { EMPTY_ARRAY }
```

Comment:
Nope, freezing provides too strong limitation.

Actually what we do wanted is not avoiding mutation per se. It can be useful in many cases like one-off service objects. Freezing would break that cases totally. We wanted to prevent leeking mutations via defaults between various instances.

In february @marshall-lee convinced me to use procs by this simple example:

```ruby
require 'dry/initializer'

class A
  extend Dry::Initializer::Mixin

  param :name, default: proc { '...' }

  def call
    name << 'HAHA'
  end
end

3.times { puts A.new.call }

# => '...HAHA'
# => '...HAHA'
# => '...HAHA'
```

In case of no-proc instantiation (was implemented from the beginning, but removed now) this would provide something like:

```ruby
class B
  extend Dry::Initializer::Mixin

  param :name, default: '...'

  def call
    name << 'HAHA'
  end
end

3.times { puts B.new.call }

# => '...HAHA'
# => '...HAHAHAHA'
# => '...HAHAHAHAHAHA'
```

With a proc we do mutation after instantiation, but prevent leaking it outside (through a single object used as default value). Now all defaults are different objects, so various instances of A become independent from each other.